### PR TITLE
fix: Fix grouping of promexporter/targetgroup

### PR DESCRIPTION
**Pull request by jiping-s**

### CODE REVIEW

File: promexporter/targetgroup.go [hunk 1]
```diff
@@ -13,40 +13,55 @@
...
+	Labels  L        `json:"labels"`
 }
 
 // GroupTargets groups targets by distinct label values
-func GroupTargets(targetList []Target) []TargetGroup {
-	targetGroupByLabels := make(map[interface{}]TargetGroup)
-
-	for _, target := range targetList {
-		if group, exists := targetGroupByLabels[target.Labels]; exists {
-			group.Targets = append(group.Targets, target.Target)
```
Comment (jiping-s) [inline promexporter/targetgroup.go:36]: This was the bug - group.Targets were assigned but the groups in the targetGroupByLabels remained unchanged.

File: generics/slices.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	initial R,
+) R {
+	var lastResult R
+	for _, item := range list {
+		lastResult = reducer(item, lastResult)
+	}
+	return lastResult
+}
+
+func GroupSlice[T any, K comparable](
```
Comment (jstaffans) [inline generics/slices.go:47]: Feels like a bit of a stretch to have this as a "generic" function, isn't it pretty much tied to the use case in promexporter?

File: promexporter/targetgroup.go (hunk 1 above)
Comment (jstaffans) [reply promexporter/targetgroup.go:36]: do you mean that the first branch of the if clause was never entered?

**Review by jstaffans — APPROVED** [review]
If you feel like it you could see what it'd look like using some open source library for generics functions like map and reduce.

File: promexporter/targetgroup.go (hunk 1 above)
Comment (jiping-s) [reply promexporter/targetgroup.go:36]: It did enter, but `group` merely holds a copy of the actual value in the `targetGroupByLabels` map, thus replacing group.Targets has no effect outside of that loop.

File: generics/slices.go (hunk 2 above)
Comment (jiping-s) [reply generics/slices.go]: removed

File: generics/slices.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
+package generics
```
[the remark this answers is not in the document]
Comment (jiping-s) [reply generics/slices.go]: replaced all generic utility functions with lo.

**Pull request merged**